prov/sm2/src/sm2_ep.c

```diff
@@ -0,0 +1,1024 @@
+/*
+ * Copyright (c) Intel Corporation. All rights reserved
+ * Copyright (c) Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <sys/uio.h>
+#include <sys/un.h>
+
+#include "ofi_hmem.h"
+#include "ofi_iov.h"
+#include "ofi_mem.h"
+#include "ofi_mr.h"
+#include "sm2.h"
+#include "sm2_fifo.h"
+#include "sm2_signal.h"
+
+#define SM2_MAX_SHUTDOWN_SECONDS 1
```

Comment:
You read that wrong, there is a goto in there and depending on what that value is set too... we will wait longer at 1 second polling intervals.

Reply:
I understand the difference between the timeout and the sleep but the use of it doesn't make any sense, especially to have a whole define for it. You're passing in a timeout of 1 and then you do a sleep only if count is less than 1 which it will only be once so it seems rather pointless for it to be an entire parameter.
It could really just be
function that processes and cleanups buffers in the fifo
if isfull return
sleep 1
call that function again
if still not full, print warning
The cleanup path in general probably needs to be redone in general because of issues like that but I think this is ok for now.

Reply:
Agree... cleanup is tricky.  The question that I am trying to answer here is how long do I want to try and wait to get all my xfer (FQE) buffers back before I shut down and zombie-fy myself.  Since I am the only one that moves my XE's from my recv queue to my free stack, if I exit without a full freestack that is guarenteeing that I will never be clean.

Reply:
Ill switch it to how you suggested.